libs/resource_partitioner/src/detail_partitioner.cpp

```diff
@@ -33,27 +32,13 @@ namespace hpx { namespace resource { namespace detail {
     HPX_NORETURN void throw_runtime_error(
         std::string const& func, std::string const& message)
     {
-        if (get_runtime_ptr() != nullptr)
-        {
-            HPX_THROW_EXCEPTION(invalid_status, func, message);
-        }
-        else
-        {
-            throw std::runtime_error(func + ": " + message);
-        }
+        HPX_THROW_EXCEPTION(invalid_status, func, message);
     }
 
     HPX_NORETURN void throw_invalid_argument(
         std::string const& func, std::string const& message)
     {
-        if (get_runtime_ptr() != nullptr)
-        {
-            HPX_THROW_EXCEPTION(bad_parameter, func, message);
-        }
-        else
-        {
-            throw std::invalid_argument(func + ": " + message);
-        }
+        HPX_THROW_EXCEPTION(bad_parameter, func, message);
```

Comment:
Does this assume to run on an HPX thread? Throwing an `hpx::exception` on a non-HPX thread is generally a bad idea... (it tries to collect a stack-backtrace by spawning a new HPX Thread).

Reply:
You're right, I didn't check this properly. I think this is okay though. Pre-runtime start we only construct a lightweight exception because the handler isn't installed at that point. During runtime, and outside of HPX threads, we construct a full/custom exception but leave out some information that only applies to HPX threads. We do get the stacktrace though, but that is guarded by this: https://github.com/STEllAR-GROUP/hpx/blob/458e59a071c7abd2330f33c10443387088566fec/src/util/backtrace/backtrace.cpp#L410-L411. If anything, the check here (partitioner) should've been with `get_self_ptr`, not `get_runtime_ptr`.

I see we don't print any of the information stored in the exception if it's thrown on a non-HPX thread, but that's orthogonal to this issue. It would certainly be convenient if we were allowed to `HPX_THROW_EXCEPTION`s anywhere, not just on HPX threads, and it seems that is the case.